File: 7-4.py
m={'+':1,'-':1,'*':2,'/':2,'(':3,')':3}
ans=[];st=[]
s=input()
i=int(0)
while i< len(s):
    #print('#')
    if s[i].isdigit() or (i==0 or s[i-1]=='(') and (s[i]=='-' or s[i]=='+'):
        str=""
        if s[i]!='+':
            str+=s[i]
        while i+1<len(s) and (s[i+1]=='.' or s[i+1].isdigit()) :
            str+=s[1+i]
            i+=1
        ans.append(str)
    else:
        if s[i]=='(':
            st.append(s[i])
        elif s[i]==')':
            while st and st[-1]!='(':
                ans.append(st[-1])
                st.pop()
            st.pop()
        else:
            while st and st[-1]!='(' and m[st[-1]] >= m[s[i]]:
                ans.append(st[-1])
                st.pop()
            st.append(s[i])
    i+=1
while st:
    ans.append(st[-1])
    st.pop()
k=0
print(*ans)

File: 7-4.cpp
#include <iostream>
#include <map>
#include <stack>
#include <string> 
using namespace std;
int main()
{
	stack<char> st;
	map<char,int> m;
	m['+']=m['-']=1;
	m['*']=m['/']=2;
	m['(']=m[')']=3;
	string s;
	cin>>s;
	int i=0,q=1;
	while(i<s.size())
	{
		if(isdigit(s[i])||((i==0)||s[i-1]=='(')&&(s[i]=='+'||s[i]=='-'))
		{
			if(q)
				q=0;
			else
				cout<<" ";
			if(s[i]=='-'||isdigit(s[i]))
				cout<<s[i];
			while(s[i+1]=='.'||isdigit(s[i+1]))
				cout<<s[++i];
			i++;
		}
		else
		{
			if(s[i]=='(')
				st.push(s[i]);
			else if(s[i]==')')
			{
				while(st.top()!='('&&!st.empty())
				{
					cout<<' '<<st.top();
					st.pop();
				}
				st.pop();//'('
			}
			else
			{
				while(!st.empty()&&st.top()!='('&&m[st.top()]>=m[s[i]])
				{
					cout<<' '<<st.top();
					st.pop();
				}
				st.push(s[i]);
			}
			i++;
		}
	}
	while(!st.empty())
	{
		cout<<' '<<st.top();
		st.pop();
	}
} 
